_posts/pl/2021-09-31-zarzadzanie-stanem-aplikacji-frontendowej.md

```diff
@@ -0,0 +1,225 @@
+
+```typescript
+const getActivitiesFeatureState = createFeatureSelector<ActivityItemsState>('activitiesState');
+
+export const getActivities = createSelector(getActivitiesFeatureState, state => state.activities);
+```
+
+Najpierw, przy pomocy funkcji `createFeatureSelector` tworzymy tzw. feature selector pozwalający na wyciągnięcie pojedynczego wycinka (_slice_) danych, nazwanego przez nas `activitiesState`. Następnie używamy go tworząc właściwy selektor przy pomocy funkcji `createSelector` i wskazując go w pierwszym jej argumencie. Drugim parametrem jest funkcja anonimowa wskazująca które dane chcemy otrzymać w wyniku działania selektora. W naszym przypadku jest to tablica czynności, o nazwie `activities`. Tak zbudowany selektor można następnie użyć w komponencie.
+
+### Spięcie całości
+Mamy już wszystkie potrzebne części logiki NgRx, z których chcemy skorzystać. Pozostaje już tylko dowiedzieć się jak ich użyć. Pokażemy sytuację, w której użytkownik klika przycisk odpowiadający za pobranie przykładowej aktywności z API, przechodząc jednocześnie przez odpowiednie miejsca w kodzie. W trakcie czytania zachęcam do spoglądania na zamieszczony wcześniej w poście diagram przepływu.
+
+1. Zacznijmy od komponentu wywołującego początkową akcję typu `Get Activity`. Budowa klasy tego komponentu może wyglądać następująco (pomijam _template_):
+```typescript
+export class ActivityApiComponent {
+
+    constructor(private store: Store<State>) {
+    }
+
+    getActivity(): void {
+        this.store.dispatch(getActivity());
+    }
+}
+```
+Jak widać w konstruktorze, wstrzykujemy obiekt Store, na którym wykonujemy metodę `dispatch` podając w jej argumencie typ akcji. Na tym kończy się odpowiedzialność komponentu.
+
+2. Wysłaną akcję przechwytuje efekt, który, zgodnie z kodem zaprezentowanym wcześniej, wysyła żądanie do API, a otrzymawszy odpowiedź przekazuje ją jako argument nowej akcji `Add Activity`.
+3. Akcja typu `Add Activity` jest przechwytywana przez reducer, który wyłuskuje z niej przekazaną z API odpowiedź i dodaje do store.
+4. W ostatniej kolejności do gry wchodzi selektor, który wykrywa zmianę store wynikającą z działania reducera. Użycie selektora w komponencie może wyglądać następująco:
+
+```typescript
+export class ActivityListComponent implements OnInit {
+
+  activities: ActivityItemModel[] = [];
+
+  constructor(private store: Store<State>) {
+  }
+
+  ngOnInit(): void {
+    this.store.select(getActivities)
+      .subscribe(activities => this.activities = activities);
+  }
+
+}
+```
+W metodzie `ngOnInit` wskazujemy, że chcemy nasłuchiwać na zmiany store przy pomocy selektora `getActivities`, a następnie przy każdej takiej zmianie aktualizujemy lokalnie utrzymywaną tablicę, którą można następnie wykorzystać w ciele komponentu.
```

Comment:
zdaje się, że ktoś już mnie kiedyś w tej kwestii edukował, ale upewnię się:
w świecie IT normalnym zwrotem jest "nasłuchiwać na"?

Reply:
Mam wrażenie, że tak, natomiast zmieniłem na "reagować"